internal/v1alpha2/controllers/bundledeployment/bundledeployment.go

```diff
@@ -0,0 +1,361 @@
+		return ctrl.Result{}, err
+	}
+
+	switch res.State {
+	case source.StateUnpackPending:
+		setUnpackStatusPending(&bundleDeployment.Status.Conditions, "pending unpack", bundleDeployment.Generation)
+		// Requeing after 5 sec for now since the average time to unpack an registry bundle locally
+		// was around ~4sec.
+		// Warning: This could end up requeing indefinitely, till an error has occurred.
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	case source.StateUnpacking:
+		setUnpackStatusPending(&bundleDeployment.Status.Conditions, "unpacking in progress", bundleDeployment.Generation)
+	case source.StateUnpacked:
+		setUnpackStatusSuccessful(&bundleDeployment.Status.Conditions, "unpacked successfully", bundleDeployment.Generation)
+	default:
+		return ctrl.Result{}, fmt.Errorf("unknown unpack state %q for bundle deployment %s: %v", res.State, bundleDeployment.GetName(), bundleDeployment.Generation)
+	}
+
+	// Unpacked contents from each source would now be available in the fs. Validate
+	// if the contents together conform to the specified format.
+	if err = b.validateContents(ctx, bundleDeployment.Spec.Format, bundledeploymentStore); err != nil {
+		validateErr := fmt.Errorf("validating contents for bundle %s with format %s: %v", bundleDeployment.Name, bundleDeployment.Spec.Format, err)
+		setValidateStatusFailing(&bundleDeployment.Status.Conditions, validateErr.Error(), bundleDeployment.Generation)
+		return ctrl.Result{}, validateErr
+	}
+	setValidateStatusSuccess(&bundleDeployment.Status.Conditions, "unpacked successfully", bundleDeployment.Generation)
+
+	// Deploy the validated contents onto the cluster.
+	// The deployer should return the list of objects which have been deployed, so that
+	// controller can be configured to set up watches for them.
+	deployResult, err := b.deployContents(ctx, bundledeploymentStore, bundleDeployment)
+	switch deployResult.State {
+	case deployer.StateIntallFailed:
+		setInstallStatusFailed(&bundleDeployment.Status.Conditions, err.Error(), bundleDeployment.Generation)
+		return ctrl.Result{}, err
+	case deployer.StateUnpgradeFailed:
+		setUnpgradeStatusFailed(&bundleDeployment.Status.Conditions, err.Error(), bundleDeployment.Generation)
+		return ctrl.Result{}, err
+	case deployer.StateReconcileFailed:
+		setReconcileStatusFailed(&bundleDeployment.Status.Conditions, err.Error(), bundleDeployment.Generation)
+		return ctrl.Result{}, err
+	case deployer.StateObjectFetchFailed:
+		setDynamicWatchFailed(&bundleDeployment.Status.Conditions, err.Error(), bundleDeployment.Generation)
+		return ctrl.Result{}, err
+	case deployer.StateDeploySuccessful:
+		setInstallStatusSuccess(&bundleDeployment.Status.Conditions, fmt.Sprintf("installed %s", bundleDeployment.GetName()), bundleDeployment.Generation)
+	default:
+		return ctrl.Result{}, fmt.Errorf("unknown deploy state %q for bundle deployment %s: %v", deployResult.State, bundleDeployment.GetName(), bundleDeployment.Generation)
+	}
```

Comment:
Is there a way we could make deployContents return the condition that should be set?